zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2508,12 +2509,18 @@ public void connect(final long maxMsToWaitUntilConnected)
    */
   public void connect(final long maxMsToWaitUntilConnected, Watcher watcher)
       throws ZkInterruptedException, ZkTimeoutException, IllegalStateException {
-    if (isClosed()) {
-      throw new IllegalStateException("ZkClient already closed!");
-    }
     boolean started = false;
-    acquireEventLock();
+
     try {
+      acquireEventLock();
+
+      if (isClosed()) {
```

Comment:
@xyuanlu can you please help me in understanding how this gonna change current gap? Here is my understanding : 
1. Current isClosed() method also tries to take a zk event lock. But after your change we would try to acquire same zk event lock "interruptibly". I don't think there is much difference between these two right? 
2. I didn't understand why did we resolve old comment about not releasing lock in case of _currentState being null.

Reply:
Thanks for the comment.
1. Having state check and operation in the same critical section is a common practice. Taking the following flow char as example. In the code logic before this change, we do acquire lock before check if state is closed and release lock after that. The actual connection and close operation is also protected by this event lock. However, the `close()` operation can happen in between of 2 release lock and acquire lock. So generally we put when doing state check and operation in the same critical section.
![Screenshot 2023-03-08 at 1 37 55 PM](https://user-images.githubusercontent.com/6392021/223856523-b686bad8-a69c-4540-9b8b-cc99f2760359.png)
2. lock.release() in the final block will be executed when try block finishes smoothly or throw exceptions. The lock is released there. 

I always find the java doc for ReentrantLock is the best place to refer to. It has the most accurate explanation and solid examples. :D (https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html#lockInterruptibly())

Reply:
Oh ok now I understand about this change! Thanks a lot @xyuanlu for explaining it patiently. Can we change the description for PR to this detailed explanation. I was wondering how are we adding extra check previously.